### ISSUE TITLE
Precompile automagically

### ISSUE THREAD

**Issue by expez** [nathanmarz/specter#26]
Pre-compilation is basically always wanted, and the only reason not to do so is because there's a cost to doing it now (more typing for new code, refactoring of old code, or lack of knowledge on the user's part) .

By turning `transform` and `select` into macros, it looks to me like the pre-compilation of navigation specs can be done automatically.

This would have several benefits:

* Easier to use, by not requiring the user to write more, or transform existing, code to get pre-compilation
* Easier to learn, by turning pre-compilation into an implementation detail and not something the user has to think about.

**Comment by nathanmarz** [comment]
I'd really like to precompile automatically, but it's not clear at all how to do that. For example:

```clojure
(defn myfn [s]
  (select [ALL :a even?] s))
```

To precompile, the [ALL :a even?] would have to be brought out of the function. That's not possible with any facilities Clojure provides.

**Comment by expez** [comment]
Why would it have to be brought out of the function?  It's right for for `select` to pre-compile, if `select` is turned into a macro.

**Comment by nathanmarz** [comment]
There's a lot of issues you'd have to deal with, such as:

- At macro-time things like ALL and filterer are just symbols. You'd have to resolve them appropriately, taking into account those symbols might be shadowed in the lexical closure.
- You need to put the compiled path somewhere. You certainly can't embed it in the outputted source as Clojure will not allow you to do that. You might be able to def it into a new global symbol and refer to that symbol within the function, but I'm not completely sure if the details will work out.

That all said, if you can get it working I'd be happy to take a pull request for it.

**Comment by tslocke** [comment]
How about simply memoizing comp-paths? With some *very* preliminary testing it seems to be working.

    (def compile-once (memoize s/comp-paths))

Then for convenience...

    (defn transform* [path f data]
      (s/compiled-transform (compile-once path) f data))

Now we can just `(transform* [ALL even?] inc data)` and it will only compile on the first call.

The problem is that if any component of the path changes, the path will be recompiled. It doesn't support parameterized paths.

We could overload `transform*` to take parameters separately:

    (defn transform*
      ([path f data]
       (s/compiled-transform (compile-once path) f data))
      ([path args f data]
       (s/compiled-transform (apply (compile-once path) args) f data)))

Now we can do things like:

    (defn score-point [data user]
      (transform* [:users keypath :score] [user] inc data))

I find it a bit burdensome to have to match the parameters with their place in the path, so I wrote some macro goodness.

    (defn extract-parameters [path]
      (loop [[x & more :as p] path
             static-path []
             args []]
        (if (seq p)
          (if (seq? x)
            (if (-> x count (= 1))
              (recur more
                     (conj static-path 'com.rpl.specter.keypath)
                     (conj args (first x)))
              (recur more
                     (conj static-path (first x))
                     (into args (rest x))))
            (recur more
                   (conj static-path x)
                   args))
          (if (-> args count pos?)
            [static-path args]
            [static-path]))))
    
    (defmacro transform [path f data]
      `(tslocke.utils/transform* ~@(extract-parameters path) ~f ~data))

With this you can do:

    (defn score-point [data user]
      (transform [:users (user) :score] inc data))

NOTE: Newly created functions always have a new hash code, so if the Specter path vector has, somewhere within it, a closure, this will thwart the memo cache. It may be that some Specter library functions do this already - I haven't checked.

If you like this idea I can submit a PR.

**Comment by vendethiel** [comment]
> At macro-time things like ALL and filterer are just symbols. You'd have to resolve them appropriately, taking into account those symbols might be shadowed in the lexical closure.

I don't think that's true, the walker already adds namespaces, so you're gonna look at a prefixed name, right?

**Comment by nathanmarz** [comment]
@vendethiel They're most definitely non-qualified symbols:

```clojure
user=> (defmacro foo [& args] (println args) (doseq [a args] (println (class a))))
#'user/foo
user=> (foo ALL srange)
(ALL srange)
clojure.lang.Symbol
clojure.lang.Symbol
```

**Comment by nathanmarz** [comment]
@tslocke Memoizing is a pretty good idea actually. It's important to recognize that the goal here is to have significantly improved common case performance as no dynamic strategy will be able to compete with explicit precompilation. 

I would also say there are two other properties we want out of a solution:

1. Consistency: The same code always performs the same and is not subject to any cache invalidation patterns that cause performance to fluctuate wildly.
2. Memory-safe: The cache size must be bounded in a reasonable way.

Because parameterized paths mean that a static number of path declarations can create effectively infinite paths at runtime, that eliminates a naive caching strategy. The only way to get #2 would be to use an eviction strategy of some sort, and that would conflict with goal #1.

I also think that a caching implementation should be completely invisible to the user and require no extra burden on their part. Specter usage should remain identical to how it is currently.

A more sophisticated strategy might be able to satisfy all these properties. If select and transform are macros, then they should be able to figure out if the path is made up of a combination of global navigators, global functions, and global navigators that require parameterization. As long as only those paths are cached, then the cache size will be bounded to the number of path declarations in the codebase. All other kinds of paths should not be cached as that would violate #2. 

So basically something like `[ALL (srange 4 9) ALL even?]` should be replaced with `((compile-once ALL srange ALL pred) 4 9 even?)`, while the following cases would not be cached:

- `[ALL some-local-symbol]`
- `(let [ALL :a] (transform [ALL even?] inc data))`

**Comment by tslocke** [comment]
> Because parameterized paths mean that a static number of path declarations can create effectively infinite paths at runtime, that eliminates a naive caching strategy.

That's already addressed in the implementation I posted. Only the static part is cached. 1 and 2 are both satisfied.

I did not satisfy this one though:

> I also think that a caching implementation should be completely invisible to the user and require no extra burden on their part. Specter usage should remain identical to how it is currently.

While this would be nice, it strikes me as hugely complicating the problem without a significant win. I don't think it will be possible, in a fully generalised way, to figure out at compile time which symbols refer to parameterised navigators. What if I call a function to return a navigator according to some logic?

Much better to have it indicated statically, in the syntax. The macro I posted just assumes that, in any form (foo a b c...), a, b and c are parameters. This could be improved upon, but it already works pretty well. I actually find it makes the path *more* readable.

Another possibility would be to wrap individual parameters in something like (? ...)

A question - what happens if I want to compile something like `(srange 3 x)` (where should be a parameter)? Can I partially apply? I.e. compile `(srange 3)` and then provide the x as a parameter? Or do I have to compile just `srange` and provide both 3 and `x` as parameters? What if I wanted `(srange x 3)`?

**Comment by nathanmarz** [comment]
I don't think your code satisfies 2.  Consider these cases:

```clojure
(defn foo [navigator data]
  (transform [ALL navigator] inc data))
```

```clojure
(defn foo [k data]
  (let [p (keypath k)]
    (transform [ALL p] inc data))
```

Each case has one declared path but at runtime can create infinite distinct paths. 

To determine whether a symbol refers to a parameterized navigator or a regular function, you'd have to resolve the var and check the value (and also check &env to ensure the symbol isn't shadowed). It would be some tricky code but I think it's possible. That said, there's always going to be a way to break it (e.g. alter-var-root), which is why I've held off on this feature. As long as the mechanisms to break it are clearly unreasonable code, it's probably OK, but this point is worth further discussion.

Parameterized navigators cannot be partially applied. You must supply all the needed parameters at once in the order that they're expected.

**Comment by tslocke** [comment]
For the second example, you would write it

    (defn foo [k data]
      (transform [ALL (keypath k)] inc data))

and it would be cached and parameterised correctly. So some burden is placed on the user to know how it works, but I would argue that's a good thing. (I actually put in a little shortcut for keypath, so you could just write `[ALL (k)]` but perhaps that's too much of a hack.)

The first example, where the entire navigator is parameterised, is more interesting. It's not entirely clear what would be wanted here. If I know my code passes one of, say, 3 different possible values, then I would probably want them all to be compiled and cached separately, which is what my solution would currently do. If, conversely, the `navigator` parameter was to be used to pass in a navigator that varied widely according to some value (i.e. something like `(keypath k)`) I would refactor to separately pass in the value to `foo`, and apply the navigator to the value inline in the call to `transform`. It's not "automagic", but it's simple to understand and easy to work with.

What would you do now with something like that first example? Can the compiler cope with that kind of path, parameterised not on a navigator parameter, but on an arbitrary choice of navigator?

How would you propose to handle that first example in your proposed solution?

Interesting discussion.

**Comment by nathanmarz** [comment]
Let's say you're new to Specter. You read through the examples to learn the basics and start using it in a project. One section of your code uses paths passed in dynamically, much like one of the examples you saw by the creator of the project in a presentation: https://github.com/nathanmarz/specter-clojure-west/blob/master/src/specter_demo/examples.clj#L316

Your project works and is performant. You test in in staging and everything works fine. Then you deploy it to production. Everything works fine for awhile, but then after a couple days your processes go OOM and die. You have a memory leak but you have no idea why. Tracking the memory leak down to the source – which is the usage of an unbounded number of dynamic paths – is going to require a hell of a debugging adventure. 

While it's true that as long as you understand the details of your proposed design, avoiding these memory leak scenarios is not that difficult. But it's unreasonable to expect all users to understand this. Heck, I could even see myself forgetting about this from time to time. Having my 2nd example above causing a memory leak is unreasonable behavior from a library. 

I wouldn't say I've officially proposed a solution to this, as I'm not convinced this can be done completely cleanly. In the macro approach I discussed, it would not cache either of my examples as both `navigator` and `p` would be in the local environment. Where the macro+caching design breaks is something like this:

```clojure
(defn foo [k data]
  (alter-var-root #'ALL (fn [_] (keypath k)))
  (transform ALL inc data)
  )
```

Now this is some very, very bad code. But still, it's surprising that it would cause a memory leak with the macro+caching design. Even this makes me uncomfortable which is why I haven't tackled this issue. I'm much more comfortable with a design in which there do not exist any usage patterns with surprising, destructive effects.

**Comment by tslocke** [comment]
How about this...

1. Have an alternate namespace, with an identical API, where the functions implement compilation caching. That way you have to make an explicit choice to use them, and can be expected to understand the consequences.

2. Raise an exception (statically, from the macro) on an attempt to cache a navigator path that contains a locally bound symbol, other than in "parameterised position" (i.e. in a seq, not in first place). 

So this would fail:

    (let [p (keypath k)]
      (transform [ALL p] inc data))

While this would pass:

    (transform [ALL (keypath k)] inc data)

the macro would remove `k` and a parameterised query would be cached.

**Comment by nathanmarz** [comment]
Gave this a lot of thought. I have some qualms about having a parallel namespace with the same operations – but my qualms with that are irrelevant because I think this is doable in a safe way via modification of the core Specter operations. Here's a design which is an evolution of both your ideas and mine:

- `defpath` and `path` become `defnav` and `nav`, respectively. `StructurePath` is renamed to `Navigator`. 
- All transform/select operations become macros that wrap the given path in a macro called `path`. Function versions of all transform/select operations should remain under different names.
- `path` statically determines whether or not to cache the path it's given or whether to compile it at runtime. `path` is its own operation so path caching can be achieved outside the context of a transform or select.
- A path is statically determined to be cached if:
  - The navigator symbols (everything in the path besides parameters) are not in the local environment. Additionally, they are resolved and determined not to be dynamic vars. 
  - All parameterized path elements  (of the form `(param-path a b c)`) are resolved to their respective vars and verified to be parameterized paths and NOT Clojure functions.
- When a path is statically determined to be cached, a unique id is statically generated for that cache site ("cache-id")
- The cache is a map from cache-id -> [uncompiled-path compiled-path] and will need to be a ConcurrentHashMap. 
- At runtime:
  - If the cache is empty for that cache id the path is dynamically compiled and the cache updated.
  - If the cache is not empty, a check is done to make sure the uncompiled-path in the cache is the same as the current path. If they are not equal, an exception is thrown (this handles the alter-var-root case I described and anything else weird that the programmer may have done).

This design guarantees that the number of cached paths is at most the number of path declarations throughout a codebase. The only ways to get two different paths for the same declaration site require doing something very strange (like alter-var-root), and the runtime checks handle those cases to give an informative exception instead of a very strange bug.

**Comment by nathanmarz** [comment]
On second thought, don't think this is going to work for cljs. Macros run in Clojure where `resolve` will not give the correct results.

**Comment by nathanmarz** [comment]
If parameterized paths are explicitly marked, then this is doable in both Clojure and ClojureScript, e.g.:

```clojure
(path ALL (<< some-path 1 2 3) :a)
```

Not sure what the best marker would be. This sacrifices the ability to use dynamic vars in paths, but I think that sacrifice is easily worth it (dynamic vars would still be usable with function versions of select/transform).

**Comment by tslocke** [comment]
I was thinking along the same lines. How about

    (? …)

Datalog uses ? to indicate parameters so there is some consistency.

BTW, is it possible to have a path parameterised on the argument to a filter function? As in

    [ALL #(> % x)]

If not, could it be achieved with something like an “apply” or “filter” navigator, that expects two parameters - a function and an argument?

**Comment by nathanmarz** [comment]
Yea, `?` would work.

`pred` is the late-bound parameterized version of using an anonymous function. So that could be redone as:

```clojure
((comp-paths ALL pred) #(> % x))
```

**Comment by tslocke** [comment]
I had offered to send a PR for this one, but the scope has rather grown since then. Is this something you expect to get to yourself?

**Comment by nathanmarz** [comment]
I'd prefer a pull request for this in the interest of having more contributors. Do you still want to take this on?

**Comment by tslocke** [comment]
I’m happy to have a crack at it. I’m in the middle of trying to get a fairly large open source project of my own finished and released, so this will have to compete with other priorities, but such is life : )

**Comment by nathanmarz** [comment]
I looked into this in detail today and it's a lot harder than it seems. My design above doesn't work – higher order navigators like `selected?` completely mess it up.

I think I have a way to do this, but it does require deep familiarity with the workings of Specter. For that reason I'm going to take this issue on myself, but I appreciate the back and forth on this issue.

**Comment by tslocke** [comment]
Great news that you’re going to have a crack at this issue. In my current project there have been a couple of places where a few lines of tricky Clojure have been reduced to a laughable easy `transform`. I can really imagine Specter becoming my default approach for all sorts of things, especially if path compilation disappears.

I’ve been working with a simple variation of the macro I posted at the top of this thread that uses the `(? …)` syntax, and it’s been perfect for my needs so far. I do remember having a problem with `selected?` but at the time I put it down to not understanding how `selected?`is used.

Would you mind posting an example of a problematic navigator, and show how you would precompile it the normal way? Thanks.

**Comment by nathanmarz** [comment]
Here's an example:

```clojure
(defn foo [k data]
  (transform [(keypath (+ 1 2)) ALL (selected? (keypath k))] inc data))
```

To precompile, you would do this:

```clojure
(let [cpath (comp-paths (keypath (+ 1 2)) ALL (selected? keypath))]
  (defn foo [k data]
     (transform (cpath k) inc data)
     ))
```

The reason it's a problem is because it makes runtime path equivalency verification too expensive, as things like `selected?` themselves run `comp-paths` on the path.

**Comment by nathanmarz** [comment]
This is now working for Clojure in the [precompiled-cached](https://github.com/nathanmarz/specter/tree/precompile-cached) branch. It does not yet work for ClojureScript as ClojureScript does not have `eval` (in most use cases), so a different, less efficient strategy will need to be taken for that platform.

Here are some notes for anyone interested:

It's completely safe to use and will never cache a path that cannot be split into a static portion and parameters (except if the user does alter-var-root on the navigators, which I think is acceptable). Note that it will NOT cache paths if a dynamic var is used for one of the navigators, which is the desired behavior.

It also will automatically factor anonymous function usage into the `pred` navigator and supply the anonymous function as a parameter at runtime.

In that branch path caching is triggered with the `com.rpl.specter.macros/ic!` macro. I will almost certainly rename that and later integrate it into select/transform. Usage is like this:

```clojure
(select (ic! :a :b :c) {:a {:b {:c 1}}})
```

For that particular use case, a nested map, the performance numbers like like this for 10,000,000 iterations on my computer:

- `get-in`: 6679.586 msecs
-  Using `ic!`: 5627.461 msecs
- Manually using `comp-paths!`: 4788.272 msecs

So, in the worst case, inline caching is ~18% slower than manual precompilation. Though for most cases, where navigation is more expensive then simply doing get operations on maps, that number will go down significantly (~2% for most of my use cases). If Clojure exposed a way to use `invokedynamic`, I believe I could get the performance to be equivalent. 

Finally, I added the ability to enforce that all paths must be cached with the `(must-cache-paths!)` function. Besides throwing an exception if a path cannot be cached, it will also print the specific reasons why.

**Comment by nathanmarz** [comment]
This is now completely implemented for 0.11.0 for both Clojure and ClojureScript. The API changes are exactly as described in https://github.com/nathanmarz/specter/issues/26#issuecomment-215229540 . Note that as a user of Specter you don't need to do anything different (such as annotate parameters) – it just works.

**Comment by tslocke** [comment]
This is great. So `transform`, `select` and many others have changed from functions to macros in this release? Are the function versions still available?

How is performance for ClojureScript? Is there much of an overhead compared to manual `comp-path`?

**Comment by vendethiel** [comment]
Sounds great. :)

**Comment by nathanmarz** [comment]
Yes, the function versions are still available in the com.rpl.specter namespace. They're called `select*`, `transform*`, etc. 

The performance for ClojureScript is excellent, though not as good as Clojure compared to manual `comp-paths`. I only tested in a node.js REPL, but for one of the most difficult cases to make fast (`[:a :b :c]` path), it's 100% slower than manual `comp-paths`. However, this is still very fast and more than adequate for the vast majority of cljs use cases. The reason that `[:a :b :c]` is difficult use case to make fast is because the navigation portion is trivial, so it greatly emphasizes any overhead that Specter adds. More complicated navigations will see the performance difference drop precipitously.
